### PR TITLE
cmd/juju/controller: allow registering public controller

### PR DESCRIPTION
**Pull request by rogpeppe**
This allows a controller with an officially signed certificate to be
registered with:

	juju register <hostname>

Because it's got an officially signed certificate, we can trust it to
obtain the controller UUID, and external user authentication lets us
know what user has been chosen without needing to select one explicitly.

This caused a significant reshuffle of the register command. In
particular, we now ask for the password before asking for
the controller name because we can't get the controller UUID
without the password.  That moves us in the direction of fixing
https://bugs.launchpad.net/juju/+bug/1614010 because we no longer prompt
for the controller name if we know that the controller cannot be stored.

We also update the controller-name input code so that it loops until a
correct controller name is supplied rather than immediately returning
an error.

The tests needed significant change because all the tests that previously
assumed that the controller name was prompted before any interaction
with the server now need a mock server.

To make this a bit more straightforward,  we add a the Prompter types
in cmd/testing which means that the interaction requirements can be
specified more declaratively.  This has the side effect of testing that
the read/write interleaving is exactly as expected.

### QA

Bootstrap a controller with an officially signed certificate and externally configured users:

    juju bootstrap --debug --config 'autocert-dns-name=<your domain name>' --config 'api-port=443' --config identity-url=https://api.jujucharms.com/identity ec2 aws

Grant everyone permission to access the controller and create models:

    juju grant everyone@external addmodel

Then configure the DNS record for your domain name to point to the IP address of the newly
bootstrapped controller.

Then, on a fresh juju client instance (or with the new controller entry removed from controller.yaml and accounts.yaml) run:

    juju register <your domain name>

That should succeed and log in as your USSO username@external.  You should then be able to create models and list models etc.

### CODE REVIEW

File: cmd/juju/controller/register.go [hunk 1]
```diff
@@ -233,74 +348,81 @@ func (c *registerCommand) maybeSetCurrentModel(ctx *cmd.Context, store jujuclien
...
 
 // getParameters gets all of the parameters required for registering, prompting
 // the user as necessary.
 func (c *registerCommand) getParameters(ctx *cmd.Context, store jujuclient.ClientStore) (*registrationParams, error) {
+	var params registrationParams
+	if strings.Contains(c.Arg, ".") || c.Arg == "localhost" {
+		// Looks like a host name - no URL-encoded base64 string should
+		// contain a dot and every public controller name should.
+		// Allow localhost for development purposes.
+		params.publicHost = c.Arg
```
Comment (reedobrien) [inline cmd/juju/controller/register.go]: can't you just return here?

File: cmd/juju/controller/register.go [hunk 2]
```diff
@@ -233,74 +348,81 @@ func (c *registerCommand) maybeSetCurrentModel(ctx *cmd.Context, store jujuclien
...
 	}
-	copy(params.key[:], info.SecretKey)
 
-	// Prompt the user for the controller name.
-	controllerName, err := c.promptControllerName(store, info.ControllerName, ctx.Stderr, ctx.Stdin)
-	if err != nil {
-		return nil, errors.Trace(err)
+	if params.publicHost != "" {
+		// No need for password shenanigans if we're using a public controller.
+		return &params, nil
```
Comment (reedobrien) [inline cmd/juju/controller/register.go]: Can't this just return up where params.publicHost is set ~line 400?

**Comment by reedobrien** [comment]
Unable to QA. I was unable to resolve the endpoints in http://paste.ubuntu.com/23080517/ to bootstrap. Updating them to be https://ec2.<region>.amazonaws.com/... resovled them but then I failed for other reasons (not authorized to perform action on one machine and `failed to bootstrap model: cannot start bootstrap instance: no "xenial" images in us-west-1 with arches [amd64]` on another.

**Comment by reedobrien** [comment]
Tried on a 3rd computer and I could bootstrap. I guess that my local DNS here is hiding the upstream changes.  I bootstrapped with a domain name I had control of, then had to add DNS pointing at the external IP of the bootstrapped controller.

Then I got a link to a paste in my browser which took me to a page, "login successful as user <myusername>".

So IIUC, QA checks out.

File: cmd/juju/controller/register.go [hunk 3]
```diff
@@ -55,27 +56,38 @@ type registerCommand struct {
...
+with a base64-encoded blob of data (the output of 'juju add-user') which
+can be copied and pasted as the <string> argument to 'register'.
+You will be prompted for a password, which, once set, causes the
 registration string to be voided. In order to start using Juju the user can now
 either add a model or wait for a model to be shared with them.  Some machine
 providers will require the user to be in possession of certain credentials in
 order to add a model.
 
+When adding a controller at a public address, authentication via some
+external third party (for example Ubuntu USSO) will be required, usually
```
Comment (rogpeppe) [inline cmd/juju/controller/register.go]: done

File: cmd/juju/controller/register.go [hunk 4]
```diff
@@ -55,27 +56,38 @@ type registerCommand struct {
...
-The user will be prompted for a password, which, once set, causes the
+The register command adds a details of a controller to the local system.
+This is done either by completing the user registration process that began
+with the 'juju add-user' command, or by providing the DNS host name
+of a public controller.
+
+To complete the user registration process, you should have been provided
+with a base64-encoded blob of data (the output of 'juju add-user') which
+can be copied and pasted as the <string> argument to 'register'.
+You will be prompted for a password, which, once set, causes the
```
Comment (rogpeppe) [inline cmd/juju/controller/register.go]: done

File: cmd/juju/controller/register.go [hunk 5]
```diff
@@ -143,67 +255,70 @@ func (c *registerCommand) Run(ctx *cmd.Context) error {
...
-		owner := names.NewUserTag(model.Owner)
-		if err := store.UpdateModel(
-			registrationParams.controllerName,
-			jujuclient.JoinOwnerModelName(owner, model.Name),
-			jujuclient.ModelDetails{model.UUID},
-		); err != nil {
-			return errors.Annotate(err, "storing model details")
+	for name, ctl := range all {
+		if ctl.ControllerUUID == controllerDetails.ControllerUUID {
+			// TODO(rogpeppe) Succeed but override the account details in this case?
```
Comment (rogpeppe) [inline cmd/juju/controller/register.go]: I've added a bug number.

File: cmd/juju/controller/register.go [hunk 6]
```diff
@@ -98,34 +110,134 @@ func (c *registerCommand) Init(args []string) error {
...
+	}
+	defer conn.Close()
+	user, ok := conn.AuthTag().(names.UserTag)
+	if !ok {
+		return errRet(errors.Errorf("logged in as %v, not a user", conn.AuthTag()))
+	}
+	return jujuclient.ControllerDetails{
+			APIEndpoints:   []string{host},
+			ControllerUUID: conn.ControllerTag().Id(),
+			// TODO(rogpeppe) it might be good to update the non-official CA cert here too
```
Comment (rogpeppe) [inline cmd/juju/controller/register.go]: I've removed the TODO. I'm not sure it's the right thing to do anyway and - it doesn't justify a bug report. (BTW only 12% of TODOs in the source code fit this format. I'm not entirely convinced it's worth having an issue for every TODO).

File: cmd/juju/controller/register.go [hunk 7]
```diff
@@ -233,74 +348,81 @@ func (c *registerCommand) maybeSetCurrentModel(ctx *cmd.Context, store jujuclien
 		if err != nil {
 			return errors.Trace(err)
 		}
-		fmt.Fprintf(ctx.Stderr, "\nCurrent model set to %q.\n\n", modelName)
-	} else {
-		fmt.Fprintf(ctx.Stderr, `
+		fmt.Fprintf(ctx.Stderr, "\nCurrent model set to %s.\n", modelName)
```
Comment (rogpeppe) [inline cmd/juju/controller/register.go]: Well, we know the model name is well formed (because it's been verified and it's a very restricted syntax) so there's no danger of ambiguity, and it looks slightly cleaner unquoted IMHO.

File: cmd/juju/controller/register.go (hunk 1 above)
Comment (rogpeppe) [inline cmd/juju/controller/register.go]: Yup, good catch - it's a hangover from when we fetched the controller name in both cases.

File: cmd/juju/controller/register.go (hunk 7 above)
Comment (axw) [inline cmd/juju/controller/register.go]: Please revert, so we are consistent with use of %q for controller names in sentences.

File: cmd/juju/controller/register.go (hunk 5 above)
Comment (axw) [inline cmd/juju/controller/register.go]: I think that would be sensible, if a little unusual. Can you please file a bug for that and reference?

File: cmd/juju/controller/register.go [hunk 8]
```diff
@@ -403,36 +525,37 @@ func (c *registerCommand) promptNewPassword(stderr io.Writer, stdin io.Reader) (
 	return password, nil
 }
 
-const errControllerConflicts = `WARNING: You already have a controller registered with the name %q. Please choose a different name for the new controller.
-
-`
-
 func (c *registerCommand) promptControllerName(store jujuclient.ClientStore, suggestedName string, stderr io.Writer, stdin io.Reader) (string, error) {
 	_, err := store.ControllerByName(suggestedName)
```
Comment (axw) [inline cmd/juju/controller/register.go]: Please skip this if suggestedName == "". It would previously always be non-empty, but that's not the case now.

File: cmd/juju/controller/register.go [hunk 9]
```diff
@@ -403,36 +525,37 @@ func (c *registerCommand) promptNewPassword(stderr io.Writer, stdin io.Reader) (
...
+		}
+		name = strings.TrimSpace(name)
+		if name == "" {
+			if suggestedName == "" {
+				fmt.Fprintln(stderr, "You must specify a non-empty controller name.")
+				continue
+			}
+			name = suggestedName
+		}
+		_, err = store.ControllerByName(name)
```
Comment (axw) [inline cmd/juju/controller/register.go]: Skip this if name == suggestedName?

File: cmd/testing/prompt.go [hunk 10]
```diff
@@ -0,0 +1,220 @@
+package testing
```
Comment (axw) [inline cmd/testing/prompt.go]: copyright

File: cmd/testing/prompt.go [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+import (
+	"io"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	gc "gopkg.in/check.v1"
+)
+
+var logger = loggo.GetLogger("cmd.testing")
```
Comment (axw) [inline cmd/testing/prompt.go]: doesn't matter much, but should start with "juju." (juju.cmd.testing)

File: cmd/testing/prompt.go [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+
+// NewSeqPrompter returns a prompter that expected prompt text matching
```
Comment (axw) [inline cmd/testing/prompt.go]: I'm having trouble parsing this. Maybe reword?

File: cmd/testing/prompt.go [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+// all the items in expectPrompts in sequence, and will reply with
+// the associated text in replies when requested.
+//
+// The expectPrompts slice may be one longer than replies,
```
Comment (axw) [inline cmd/testing/prompt.go]: this is out of date, there's no expectPrompts

File: cmd/testing/prompt.go [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+// that will not occur in the rest of the text) and runs to the end of a
+// line.
+//
+// All output text in between user input is treated as regular expressions.
+//
+// As a special case, if an input marker is followed only by a single input
+// marker on that line, the checker will cause io.EOF to be returned for
+// that prompt.
+//
+// The returned interactionChecker wraps a Prompter and checks that each
```
Comment (axw) [inline cmd/testing/prompt.go]: s/interactionChecker/SeqPrompter/? and below

File: cmd/testing/prompt.go [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+// read and write corresponds to the expected action in the sequence.
+//
+// After all interaction is done, CheckDone or AssertDone should be closed to
```
Comment (axw) [inline cmd/testing/prompt.go]: s/closed/called/

File: cmd/juju/controller/register.go [hunk 16]
```diff
@@ -98,34 +110,134 @@ func (c *registerCommand) Init(args []string) error {
...
 	}
 
+	fmt.Fprintf(
+		ctx.Stderr, "\nWelcome, %s. You are now logged into %q.\n",
+		friendlyUserName(accountDetails.User), controllerName,
+	)
+	return c.maybeSetCurrentModel(ctx, store, controllerName, accountDetails.User, models)
+}
+
+func friendlyUserName(user string) string {
```
Comment (rogpeppe) [inline cmd/juju/controller/register.go]: Yay!

File: cmd/testing/prompt.go (hunk 10 above)
Comment (rogpeppe) [inline cmd/testing/prompt.go]: done

File: cmd/testing/prompt.go (hunk 11 above)
Comment (rogpeppe) [inline cmd/testing/prompt.go]: done

File: cmd/testing/prompt.go (hunk 13 above)
Comment (rogpeppe) [inline cmd/testing/prompt.go]: This whole comment is a relic. Deleted.

File: cmd/testing/prompt.go (hunk 15 above)
Comment (rogpeppe) [inline cmd/testing/prompt.go]: done

File: cmd/juju/controller/register.go [hunk 17]
```diff
@@ -55,27 +56,38 @@ type registerCommand struct {
...
-Registers a Juju user to a controller.`[1:]
+Registers a controller.`[1:]
 
 var usageRegisterDetails = `
-Connects to a controller and completes the user registration process that began
-with the `[1:] + "`juju add-user`" + ` command. The latter prints out the 'string' that is
-referred to in Usage.
-
-The user will be prompted for a password, which, once set, causes the
+The register command adds a details of a controller to the local system.
```
Comment (rogpeppe) [inline cmd/juju/controller/register.go]: done

File: cmd/juju/controller/register.go [hunk 18]
```diff
@@ -98,34 +110,134 @@ func (c *registerCommand) Init(args []string) error {
...
+	return c.nonPublicControllerDetails(ctx, p)
+}
+
+// publicControllerDetails returns controller and account details to be registered
+// for the given public controller host name.
+func (c *registerCommand) publicControllerDetails(host string) (jujuclient.ControllerDetails, jujuclient.AccountDetails, error) {
+	errRet := func(err error) (jujuclient.ControllerDetails, jujuclient.AccountDetails, error) {
+		return jujuclient.ControllerDetails{}, jujuclient.AccountDetails{}, err
+	}
+	if !strings.Contains(host, ":") {
```
Comment (rogpeppe) [inline cmd/juju/controller/register.go]: *shrug* done

File: cmd/juju/controller/register.go (hunk 5 above)
Comment (rogpeppe) [inline cmd/juju/controller/register.go]: Done (bug was already filed)

File: cmd/juju/controller/register.go (hunk 8 above)
Comment (rogpeppe) [inline cmd/juju/controller/register.go]: done

File: cmd/juju/controller/register.go (hunk 9 above)
Comment (rogpeppe) [inline cmd/juju/controller/register.go]: I think I prefer to keep the logic simpler - it does no harm to check in all cases.

File: cmd/testing/prompt.go (hunk 14 above)
Comment (rogpeppe) [inline cmd/testing/prompt.go]: done

**Comment by rogpeppe** [comment]
$$merge$$
